packages/runtime/container-runtime/src/dataStoreContext.ts

```diff
@@ -238,10 +238,14 @@ export abstract class FluidDataStoreContext extends TypedEventEmitter<IFluidData
 
         const thisSummarizeInternal =
             async (fullTree: boolean, trackState: boolean) => this.summarizeInternal(fullTree, trackState);
+
+        // Add self route (empty string) to used routes in the summarizer node. If GC is enabled, the used routes will
+        // be updated as per the GC data.
         this.summarizerNode = createSummarizerNode(
             thisSummarizeInternal,
             async () => this.getGCDataInternal(),
-            async () => this.getInitialGCData(),
+            async () => this.getInitialGCDetails(),
+            [""] /* usedRoutes */,
```

Comment:
So after reading the code, one part is confusing me: why do we need `usedRoutes` passed into createSummarizerNode (create child) calls?

At first I was thinking it would be needed to get the initial value of `referenceUsedRoutes`, but that actually occurs via the initial details promise (which seems correct).

Is it only there to update the `wipSerializedUsedRoutes` when creating a child during a summarize call?

And sort of a part II to that question: when will `usedRoutes` get set, other than the constructor? - Is this part of the next PR?

It feels like we must calculate `wipSerializedUsedRoutes` during summarize every time anyway, so why do we need to pass it in the constructor?

Reply:
usedRoutes will be set every time GC is run. And GC will be run before summarize happens. I am adding one part of that change here in the next iteration.
As to why usedRoutes needs to be passed to createChild. Consider the following scenario:
- A data store is not yet realized and it has couple of DDSs.
- We run GC and calculate usedRoutes. The data store's usedRoutes are updated but since the children are not yet created, we can't update thier usedRoutes yet.
- We start summarize and the data store is realized (because something changed and we cannot reuse previous summary). - We create the channel contexts and the child summarizer nodes. At this point, we need to update the usedRoutes in the child summarizer nodes with the usedRoutes that was generated during GC. This is needed for 2 reasons:
  1. The child's wipSerializedUsedRoutes needs to be updated.
  2. If the child has to summarized again, the usedRoutes has to be added to the GC blob in the summary.

Reply:
That makes sense, thanks. I think I was losing a little bit of the story since I was missing seeing the code where usedRoutes actually gets set.

Reply:
Okay so when creating any child `SummarizerNodeWithGC`, if we are _not_ currently in the middle of summarizing, then we have no need to pass any useful information in the usedRoutes, except for that we want to make sure if GC is disabled for this run that we don't accidentally mark all nodes as unreferenced.

That means that we only care about usedRoutes passed to createChild when in the middle of a summary to set the wipSerializedUsedRoutes (which is what gets added to the GC blob in the summary, right?).  I didn't realize this until now, but it feels like maybe the code could be simplified and the usedRoutes parameter removed from the public createChild if `SummarizerNodeWithGC` knew how to pass the usedRoutes to its child.

I think you mentioned this in our discussion, and the hesitation was that the logic for going from parent node routes to child node routes may not belong in `SummarizerNodeWithGC`. If that's the case, can we just inject that logic somehow? And since for now it _is_ consistent, we could just put it there. I am talking about the logic here: `DataStores.updateUsedRoutes`.

Reply:
Above is just a thought, I definitely won't hold the PR on this. Either way, I think your answer to this question will help solidify my understanding regardless of which path you choose :)

Reply:
Technically, if we have usedRoutes that were generated by GC,  we need to pass them to the child when its created regardless of whether summarize in progress or not. This will ensure that the next time summarize happens, the child has its usedRoutes and we are not relying on usedRoutes to be generated again before the next summarize.

Yes, the summarizer node can look at its usedRoutes, find which belongs to a child when creating and pass it along bypassing this step in the data store. The only hesitation that I have is the one you pointed out - summarizer node does not understand used routes since the route there can be decoded only by the data store / channel context. We could abstract the logic in `DataStores.updateUsedRoutes` and have the summarizer node call it.
The other problem is that today the summarizer node's id matches with the data store / channel context's id. But if that stops being true, then it would have to call into its creator, pass in the child's id and ask for the child's used routes.

I am not sure which approach is better. In both, we will have to add a parameter to createChild and SummarizerNodeWithGC's constructor.

Reply:
Also, the usedRoutes that is added to the GC blob in summary is SummarizerNodeWithGC.usedRoutes and not SummarizerNodeWithGC.wipSerializedUsedRoutes. That logic is in `summarizeInternal` method of DataStoreContext.